File: legacy-cpp-backend/delivery.cpp
#include "delivery.h"

#include <algorithm>
#include <functional>
#include <limits>
#include <queue>

DeliveryManager g_delivery_manager;

void DeliveryManager::addUndirectedEdge(int u, int v, int w) {
  adj_[u].push_back({v, w});
  adj_[v].push_back({u, w});
}

void DeliveryManager::initialize() {
  locations_ = {
      {0, "Cafe", "", 1},
      {1, "Main Entrance / Ground Floor Corridor", "General", 0},
      {2, "Staircase Left", "General", 0},
      {3, "Staircase Right", "General", 0},
      {4, "Floor 0 - Block A (Left Wing)", "General", 0},
      {5, "Floor 0 - Block B (Center)", "General", 0},
      {6, "Floor 0 - Block C (PMSE Dept)", "PMSE", 0},
      {7, "Floor 1 - Block A (CSE Dept)", "CSE", 1},
      {8, "Floor 1 - Block B (ECE Dept)", "ECE", 1},
      {9, "Floor 1 - Block C (Right Wing)", "General", 1},
      {10, "Computer Lab (Ground Floor)", "Lab", 0},
      {11, "Electronics Lab (Ground Floor)", "Lab", 0},
      {12, "Faculty Lounge (Floor 1)", "General", 1},
  };

  adj_.assign(locations_.size(), {});

  addUndirectedEdge(0, 8, 10);
  addUndirectedEdge(0, 12, 15);
  addUndirectedEdge(0, 3, 20);
  addUndirectedEdge(1, 4, 40);
  addUndirectedEdge(1, 5, 20);
  addUndirectedEdge(1, 6, 50);
  addUndirectedEdge(4, 2, 25);
  addUndirectedEdge(5, 2, 30);
  addUndirectedEdge(5, 3, 30);
  addUndirectedEdge(6, 3, 25);
  addUndirectedEdge(4, 10, 35);
  addUndirectedEdge(5, 11, 40);
  addUndirectedEdge(2, 7, 40);
  addUndirectedEdge(2, 9, 50);
  addUndirectedEdge(3, 8, 40);
  addUndirectedEdge(3, 9, 35);
  addUndirectedEdge(7, 8, 30);
  addUndirectedEdge(8, 9, 30);
  addUndirectedEdge(7, 12, 20);
  addUndirectedEdge(8, 12, 20);
}

DeliveryResult DeliveryManager::findShortestPath(int destination) const {
  const int n = static_cast<int>(locations_.size());
  const int source = 0;
  const int INF = std::numeric_limits<int>::max();

  std::vector<int> dist(n, INF);
  std::vector<int> parent(n, -1);
  std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>,
                      std::greater<std::pair<int, int>>>
      pq;

  dist[source] = 0;
  pq.push({0, source});

  while (!pq.empty()) {
    auto current = pq.top();
    pq.pop();
    int d = current.first;
    int node = current.second;

    if (d > dist[node]) {
      continue;
    }

    for (const auto& edge : adj_[node]) {
      int next = edge.first;
      int w = edge.second;
      if (dist[node] != INF && dist[node] + w < dist[next]) {
        dist[next] = dist[node] + w;
        parent[next] = node;
        pq.push({dist[next], next});
      }
    }
  }

  if (destination < 0 || destination >= n || dist[destination] == INF) {
    return {false, -1, {}};
  }

  std::vector<int> reversed_path;
  int cur = destination;
  while (cur != -1) {
    reversed_path.push_back(cur);
    cur = parent[cur];
  }
  std::reverse(reversed_path.begin(), reversed_path.end());

  std::vector<std::string> path_names;
  for (int idx : reversed_path) {
    path_names.push_back(locations_[idx].name);
  }

  return {true, dist[destination], path_names};
}

std::vector<LocationInfo> DeliveryManager::getLocations() const { return locations_; }

std::vector<std::string> DeliveryManager::getAdjacencyListForDisplay() const {
  std::vector<std::string> lines;
  for (int u = 0; u < static_cast<int>(adj_.size()); ++u) {
    std::string line = locations_[u].name + " -> ";
    for (size_t i = 0; i < adj_[u].size(); ++i) {
      int v = adj_[u][i].first;
      int w = adj_[u][i].second;
      line += locations_[v].name + " (" + std::to_string(w) + "m)";
      if (i + 1 < adj_[u].size()) {
        line += ", ";
      }
    }
    lines.push_back(line);
  }
  return lines;
}

File: lib/delivery.ts
// Direct port of DeliveryManager from delivery.cpp.
// Models the building as a weighted undirected graph (corridors,
// staircases, department blocks) and runs Dijkstra's algorithm with
// a binary min-heap to find the shortest walking distance — in
// metres — from the cafe to any department, lab, or lounge.

import { MinHeap } from "./minHeap";
import type { DeliveryResult, LocationInfo } from "./types";

const LOCATIONS: LocationInfo[] = [
  { id: 0, name: "Cafe", department: "", floor: 1 },
  { id: 1, name: "Main Entrance / Ground Floor Corridor", department: "General", floor: 0 },
  { id: 2, name: "Staircase Left", department: "General", floor: 0 },
  { id: 3, name: "Staircase Right", department: "General", floor: 0 },
  { id: 4, name: "Floor 0 - Block A (Left Wing)", department: "General", floor: 0 },
  { id: 5, name: "Floor 0 - Block B (Center)", department: "General", floor: 0 },
  { id: 6, name: "Floor 0 - Block C (PMSE Dept)", department: "PMSE", floor: 0 },
  { id: 7, name: "Floor 1 - Block A (CSE Dept)", department: "CSE", floor: 1 },
  { id: 8, name: "Floor 1 - Block B (ECE Dept)", department: "ECE", floor: 1 },
  { id: 9, name: "Floor 1 - Block C (Right Wing)", department: "General", floor: 1 },
  { id: 10, name: "Computer Lab (Ground Floor)", department: "Lab", floor: 0 },
  { id: 11, name: "Electronics Lab (Ground Floor)", department: "Lab", floor: 0 },
  { id: 12, name: "Faculty Lounge (Floor 1)", department: "General", floor: 1 },
];

const EDGES: Array<[number, number, number]> = [
  [0, 8, 10],
  [0, 12, 15],
  [0, 3, 20],
  [1, 4, 40],
  [1, 5, 20],
  [1, 6, 50],
  [4, 2, 25],
  [5, 2, 30],
  [5, 3, 30],
  [6, 3, 25],
  [4, 10, 35],
  [5, 11, 40],
  [2, 7, 40],
  [2, 9, 50],
  [3, 8, 40],
  [3, 9, 35],
  [7, 8, 30],
  [8, 9, 30],
  [7, 12, 20],
  [8, 12, 20],
];

export class DeliveryManager {
  private adjacency: Array<Array<[number, number]>> = [];

  constructor() {
    this.adjacency = LOCATIONS.map(() => []);
    for (const [u, v, w] of EDGES) {
      this.adjacency[u].push([v, w]);
      this.adjacency[v].push([u, w]);
    }
  }

  getLocations(): LocationInfo[] {
    return LOCATIONS;
  }

  getAdjacencyListForDisplay(): string[] {
    return this.adjacency.map((edges, u) => {
      const parts = edges.map(
        ([v, w]) => `${LOCATIONS[v].name} (${w}m)`
      );
      return `${LOCATIONS[u].name} -> ${parts.join(", ")}`;
    });
  }

  findShortestPath(destination: number): DeliveryResult {
    const n = LOCATIONS.length;
    const source = 0;
    const INF = Number.MAX_SAFE_INTEGER;

    const dist = new Array<number>(n).fill(INF);
    const parent = new Array<number>(n).fill(-1);
    const visited = new Array<boolean>(n).fill(false);

    const heap = new MinHeap<[number, number]>((a, b) => a[0] < b[0]);
    dist[source] = 0;
    heap.push([0, source]);

    while (!heap.isEmpty()) {
      const [d, node] = heap.pop()!;
      if (visited[node]) continue;
      if (d > dist[node]) continue;
      visited[node] = true;

      for (const [next, w] of this.adjacency[node]) {
        if (dist[node] + w < dist[next]) {
          dist[next] = dist[node] + w;
          parent[next] = node;
          heap.push([dist[next], next]);
        }
      }
    }

    if (destination < 0 || destination >= n || dist[destination] === INF) {
      return { reachable: false, distance: -1, path: [], steps: [] };
    }

    const reversedPath: number[] = [];
    let cur = destination;
    while (cur !== -1) {
      reversedPath.push(cur);
      cur = parent[cur];
    }
    reversedPath.reverse();

    return {
      reachable: true,
      distance: dist[destination],
      path: reversedPath.map((idx) => LOCATIONS[idx].name),
      steps: reversedPath.map((idx) => ({
        name: LOCATIONS[idx].name,
        distanceFromCafe: dist[idx],
      })),
    };
  }
}
